_posts/2025-05-05-plan-execute-agent.md

```diff
@@ -0,0 +1,447 @@
+
+We'll demonstrate how to use the plan-execute-reflect agent to build an observability agent capable of root-causing an issue in a microservices application. We'll show you how to register the agent, run a task, and interpret the output, using natural language input with minimal setup.
+
+We'll simulate a failure in a microservices environment: a cart failure in an e-commerce application. For this, we'll use the OpenTelemetry demo application, a reference application that mimics a real-world system.
+
+This application emits telemetry data, including logs, traces, and metrics, to OpenSearch, giving us rich observability signals. We'll trigger a cart failure, then ask our observability agent to diagnose the root cause using these signals.
+
+For more information about the demo application, see [OpenTelemetry-Demo](https://opentelemetry.io/docs/demo/).
+
+### Triggering a cart failure
+
+The OpenTelemetry Demo uses `flagd`, a simple feature-flag service, to manage built-in fault scenarios. You can toggle flag values using a web UI or by editing the demo's configuration file directly. To simulate a cart failure, perform the following steps:
+
+1. Deploy the OpenTelemetry Demo application and navigate to the feature flag dashboard at[http://localhost:8080/feature](http://localhost:8080/feature).
+2. Find **cartFailure** and toggle it to the **On** position, then select **Save**.
+3. To access the cart, go to [http://localhost:8080/cart](http://localhost:8080/cart). Once the flag is enabled, the **Empty Cart** button in the frontend UI will stop working. 
+4. To generate the error condition, add at least one item to your shopping cart and then select the **Empty Cart** button.
+
+The operation will fail, triggering error spans and logs in OpenSearch corresponding to the Cart Service failures. 
+
+With the cart failure now active, the demo application will emit error-level spans and log entries for the Cart Service. These signals form the basis of our root cause analysis.
+
+## Investigating the issue using an observability agent
+
+Let's now try to identify the root cause of this issue using OpenSearch's new plan-execute-reflect agent. Instead of manually querying logs or sifting through traces, we'll describe the problem to the agent and let it do the investigation. Let's set up the agent using the following steps.
```

Comment:
```suggestion
Let's now try to identify the root cause of this issue using OpenSearch's new plan-execute-reflect agent. Instead of manually querying logs or sifting through traces, we'll describe the problem to the agent and let it perform the investigation. Let's set up the agent using the following steps.
```